### PR TITLE
[Store] Add SSD free-ratio-first allocation strategy

### LINKED ISSUE #2449: [Feature Request]: SSD free-ratio-first allocation strategy for load-balanced SSD offload
## 1. Problem

When SSD offload is enabled, existing Mooncake Store allocation strategies do not account for per-segment SSD utilization. As a result, allocation may still prefer segments whose SSD space is already close to full, while other segments with more SSD headroom remain underutilized.

This can lead to uneven SSD usage, earlier SSD overflow / eviction pressure on hot segments, and inefficient use of cluster-wide SSD capacity.

## 2. Proposal

Add a new allocation strategy:

```text
ssd_free_ratio_first
```

The strategy ranks candidate segments by SSD free ratio:

```text
ssd_free_ratio = (ssd_total_bytes - ssd_used_bytes) / ssd_total_bytes
```

Segments with higher SSD free ratio should be preferred during allocation, while preserving existing allocation constraints such as preferred segments, excluded segments, multi-replica placement, and full allocator handling.

### Goals

* Add a new `ssd_free_ratio_first` allocation strategy.
* Expose per-segment SSD capacity and usage through an `SsdMetricsProvider` interface.
* Track SSD usage in `LocalDiskSegment`.
* Support configuration through:

```bash
./mooncake_master --allocation_strategy=ssd_free_ratio_first
```

* Add unit tests and documentation.

### Non-Goals

* Change the behavior of existing strategies such as `random`, `free_ratio_first`, or `cxl`.
* Rebalance or migrate already-placed objects across SSDs.
* Implement cross-node SSD migration.

## 4. Compatibility

The new strategy should be opt-in.

Existing allocation strategies should remain unchanged. If SSD metrics are unavailable, the new strategy should fall back gracefully to the existing base allocation behavior.

## 5. Acceptance Criteria

* `ssd_free_ratio_first` can be selected through `--allocation_strategy`.
* Candidate segments are ranked by SSD free ratio when SSD metrics are available.
* Allocation still respects preferred / excluded segments and multi-replica constraints.
* Existing allocation strategies continue to pass their current tests.
* New tests cover:

  * normal SSD-aware allocation
  * fallback behavior without SSD metrics
  * full allocator / edge cases
  * load-balancing behavior across segments with different SSD free ratios

## 6. Related Implementation Notes

A possible implementation may include:

* `SsdMetricsProvider`
* `SsdFreeRatioFirstAllocationStrategy`
* `ssd_used_bytes` tracking in `LocalDiskSegment`
* documentation under `docs/source/design/ssd-free-ratio-first-allocation.md`

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues and read the [documentation](https://kvcache-ai.github.io/Mooncake/)

### PR DESCRIPTION
**Pull request by zchuango**
## Description

Closes #2449 

This PR adds a new **SSD free-ratio-first allocation strategy** (`ssd_free_ratio_first`) to Mooncake Store's master. When SSD offload is enabled, existing allocation strategies (random, free_ratio_first) are DDR-only and do not consider SSD utilization. This can cause disproportionate allocation traffic, leading to SSD overflow on some segments while others remain underutilized.

### How it works

`SsdFreeRatioFirstAllocationStrategy` ranks candidate segments by their **SSD free ratio** (`(total - used) / total`) and preferentially allocates to segments with more available SSD space. This naturally balances load across segments based on their actual SSD headroom.

### Key changes

- **`SsdMetricsProvider` interface**: Abstract interface for querying SSD capacity and usage per segment
- **`SsdFreeRatioFirstAllocationStrategy`**: New allocation strategy with SSD-ratio-based candidate ranking
- **SSD usage tracking**: `ssd_used_bytes` atomic counter in `LocalDiskSegment`, updated on offload success and disk replica eviction
- **Configuration**: `--allocation_strategy=ssd_free_ratio_first` gflag
- **Design document**: `docs/source/design/ssd-free-ratio-first-allocation.md`

### Usage

```bash
./mooncake_master --allocation_strategy=ssd_free_ratio_first
```

## Module

- [ ] Transfer Engine (`mooncake-transfer-engine`)
- [x] Mooncake Store (`mooncake-store`)
- [ ] Mooncake EP (`mooncake-ep`)
- [ ] Integration (`mooncake-integration`)
- [ ] P2P Store (`mooncake-p2p-store`)
- [ ] Python Wheel (`mooncake-wheel`)
- [ ] PyTorch Backend (`mooncake-pg`)
- [ ] Mooncake RL (`mooncake-rl`)
- [ ] CI/CD
- [x] Docs
- [ ] Other

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other

## How Has This Been Tested?

- gtest unit tests in `mooncake-store/tests/allocation_strategy_test.cpp`:
  - `SSD_FREE_RATIO_FIRST` added to parameterized test suite, covering basic allocation, preferred/excluded segments, multi-replica, full allocators, and edge cases
  - `SsdFreeRatioFirstLoadBalancingDistribution`: verifies segments with more SSD free space receive more allocations
  - `SsdFreeRatioFirstWithoutMetricsProviderAllocates`: verifies graceful fallback when no SSD metrics are available
- Existing allocation strategies (random, free_ratio_first, cxl) are unaffected: the new `Allocate` overload delegates to the base version by default

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have formatted my own code using `./scripts/code_format.sh` before submitting.
- [x] I have updated the documentation.
- [x] I have added tests to prove my changes are effective.

### CODE REVIEW

**Comment by zchuango** [comment]
Hi @ykwd @stmatengss @YiXR @XucSh @ShangmingCai , sorry for the ping. This PR adds the `ssd_free_ratio_first` allocation strategy for Mooncake Store, with tests and documentation included.

The previous bot comments have been addressed. Could you please take a look when you have time? Thanks!

**Comment by LujhCoconut** [comment]
It looks like the comments from gemini-code-assist haven't been fully addressed yet.

1.  PR only updates `ssd_used_bytes` in two paths: `NotifyOffloadSuccess` (+increment) and `EvictDiskReplica` (-decrement). However, LOCAL_DISK replicas can be destroyed through other paths without decrementing the counter (as noted in Gemini's Comments#1).

2.  Gemini's Comments#2 on clamping to [0, total] before calculating the free ratio should also be addressed. `ssd_used_bytes` is updated with `memory_order_relaxed`. During concurrent `fetch_add`/`fetch_sub`, `used` may transiently exceed `total` or go negative, producing an invalid free ratio (negative or >1.0).

File: mooncake-store/src/master_service.cpp [hunk 1]
```diff
@@ -1407,9 +1407,13 @@ auto MasterService::AllocateAndInsertMetadata(
...
         }
 
+        ScopedLocalDiskSegmentAccess ssd_access =
+            segment_manager_.getLocalDiskSegmentAccess();
+
         auto allocation_result = allocation_strategy_->Allocate(
             allocator_manager, value_length, config.replica_num,
-            preferred_segments);
+            preferred_segments, std::set<std::string>(), ReplicaType::MEMORY,
+            &ssd_access);
```
Comment (LujhCoconut) [inline mooncake-store/src/master_service.cpp]: Just a suggestion — non-blocking for the merge. The examples provided are for reference only.

`AllocateAndInsertMetadata` always creates a `ScopedLocalDiskSegmentAccess` and passes it to the allocation strategy, even when the active strategy is not SSD-aware. The non-SSD strategies ignore the provider parameter, so the allocation is correct, but the RAII guard acquisition is unnecessary overhead.

Suggested 
```cpp
const SsdMetricsProvider* ssd_provider = nullptr;
ScopedLocalDiskSegmentAccess ssd_access;  // default-constructed, no lock
if (allocation_strategy_type_ == AllocationStrategyType::SSD_FREE_RATIO_FIRST) {
    ssd_access = segment_manager_.getLocalDiskSegmentAccess();
    ssd_provider = &ssd_access;
}
auto allocation_result = allocation_strategy_->Allocate(
    allocator_manager, value_length, config.replica_num,
    preferred_segments, std::set<std::string>(), ReplicaType::MEMORY,
    ssd_provider);
```

**Comment by LujhCoconut** [comment]
> It looks like the comments from gemini-code-assist haven't been fully addressed yet.
> 
> 1. PR only updates `ssd_used_bytes` in two paths: `NotifyOffloadSuccess` (+increment) and `EvictDiskReplica` (-decrement). However, LOCAL_DISK replicas can be destroyed through other paths without decrementing the counter (as noted in Gemini's Comments#1).
> 2. Gemini's Comments#2 on clamping to [0, total] before calculating the free ratio should also be addressed. `ssd_used_bytes` is updated with `memory_order_relaxed`. During concurrent `fetch_add`/`fetch_sub`, `used` may transiently exceed `total` or go negative, producing an invalid free ratio (negative or >1.0).

Suggestion for issue#1:

Decrement `ssd_used_bytes` centrally when LOCAL_DISK replicas are destroyed, rather than at every call site. Two options:

- **Option A (recommended)**: Make `ObjectMetadata` destructor (or `EraseReplicas`) aware of LOCAL_DISK replica cleanup. When replicas are destroyed, check if they are LOCAL_DISK and decrement the corresponding segment's `ssd_used_bytes`. This requires passing a reference to the segment manager or a callback into the metadata layer.
- **Option B**: Replace all `EraseReplicas` calls that may touch LOCAL_DISK replicas with `PopReplicas` (which returns the erased replicas), then manually decrement at each call site. More invasive and error-prone.

**Comment by zchuango** [comment]
> > It looks like the comments from gemini-code-assist haven't been fully addressed yet.
> > 
> > 1. PR only updates `ssd_used_bytes` in two paths: `NotifyOffloadSuccess` (+increment) and `EvictDiskReplica` (-decrement). However, LOCAL_DISK replicas can be destroyed through other paths without decrementing the counter (as noted in Gemini's Comments#1).
> > 2. Gemini's Comments#2 on clamping to [0, total] before calculating the free ratio should also be addressed. `ssd_used_bytes` is updated with `memory_order_relaxed`. During concurrent `fetch_add`/`fetch_sub`, `used` may transiently exceed `total` or go negative, producing an invalid free ratio (negative or >1.0).
> 
> Suggestion for issue#1:
> 
> Decrement `ssd_used_bytes` centrally when LOCAL_DISK replicas are destroyed, rather than at every call site. Two options:
> 
> * **Option A (recommended)**: Make `ObjectMetadata` destructor (or `EraseReplicas`) aware of LOCAL_DISK replica cleanup. When replicas are destroyed, check if they are LOCAL_DISK and decrement the corresponding segment's `ssd_used_bytes`. This requires passing a reference to the segment manager or a callback into the metadata layer.
> * **Option B**: Replace all `EraseReplicas` calls that may touch LOCAL_DISK replicas with `PopReplicas` (which returns the erased replicas), then manually decrement at each call site. More invasive and error-prone.

Thanks for the detailed review.

The two Gemini comments have been addressed.

For the SSD usage accounting issue, we did not choose Option A because making `ObjectMetadata` / the metadata layer aware of SSD usage accounting would mix metadata management with segment-level resource accounting. Instead, the cleanup is handled on the `MasterService` side through centralized helper logic: when `LOCAL_DISK` replicas are erased or metadata is removed, the removed replicas are collected and the corresponding segment `ssd_used_bytes` is updated there. This avoids scattering manual decrement logic at every call site while keeping the metadata layer focused on metadata only.

For the free-ratio calculation issue, `used` is now clamped to `[0, total]` before calculating the SSD free ratio, so transient concurrent counter values will not produce a negative ratio or a ratio greater than 1.0.

For the non-blocking suggestion about unnecessary `ScopedLocalDiskSegmentAccess` acquisition, I agree with the concern. I plan to use lazy construction with `std::optional<ScopedLocalDiskSegmentAccess>`, so the local-disk access guard is only acquired when the active strategy is `SSD_FREE_RATIO_FIRST`; other strategies will pass `nullptr` and avoid the extra overhead. Still working on that.

File: mooncake-store/src/master_service.cpp [hunk 2]
```diff
@@ -3238,6 +3284,25 @@ auto MasterService::NotifyOffloadSuccess(
...
+                segment_manager_.getLocalDiskSegmentAccess();
+            auto& client_segments = ssd_access.getClientLocalDiskSegment();
+            auto disk_it = client_segments.find(client_id);
+            if (disk_it != client_segments.end()) {
+                disk_it->second->ssd_used_bytes.fetch_add(
+                    total_ssd_increment, std::memory_order_relaxed);
+            }
+        }
+    }
+
```
Comment (LujhCoconut) [inline mooncake-store/src/master_service.cpp:4380]: PR added the following block (`// Track SSD usage for the client`) after the for loop in `NotifyOffloadSuccess`. This causes `ssd_used_bytes` to diverge from the actual LOCAL_DISK replicas in metadata, leading the SSD free-ratio allocation strategy to rank segments based on incorrect information and eventually lose its load-balancing effect

**Problem**
In the existing for loop, `AddReplica` may partially succeed:
- Non-`OBJECT_NOT_FOUND` error → early return, new block never executes → successful task's replica in metadata but `ssd_used_bytes` not incremented (under-count)
- `OBJECT_NOT_FOUND` (object deleted) → not fatal, continues → new block accumulates all tasks' `data_size` → no replica in metadata to ever decrement (over-count / leak)

**Suggested fix**
Delete the new block, move `fetch_add` inside loop after each successful `AddReplica`

```cpp
       // `has_disk` and `disk_it` should be initialized before the loop）
       auto disk_it = client_segments.find(client_id);                                                                   
       bool has_disk = (disk_it != client_segments.end());

        for (size_t i = 0; i < tasks.size(); ++i) {
        .......
        auto res = AddReplica(client_id, object_id.user_key,
                              object_id.tenant_id, replica);
        if (!res) {
            if (res.error() == ErrorCode::OBJECT_NOT_FOUND) {
                continue;  // Not in metadata, skip 
            }
            return tl::make_unexpected(res.error());
        }

        // New: accumulate SSD usage only after successful AddReplica
        if (has_disk && metadata.data_size > 0) {
            disk_it->second->ssd_used_bytes.fetch_add(
                metadata.data_size, std::memory_order_relaxed);
        }
    }
    // Delete the "Track SSD usage for the client" block outside the loop
```

**Comment by LujhCoconut** [comment]
Thanks for this PR and your patience. Apart from the comments above, one final note: please remember to sync `docs/source/design/ssd-free-ratio-first-allocation.md`.

**Comment by zchuango** [comment]
> Thanks for this PR and your patience. Apart from the comments above, one final note: please remember to sync `docs/source/design/ssd-free-ratio-first-allocation.md`.

I have updated [mooncake-store/src/master_service.cpp](https://github.com/kvcache-ai/Mooncake/pull/2450/files/3bcb3320ffc6c9fdb194e066337ecccbe1cba184#diff-d51d7b6620a89960e72bc57fcdedbc78331f1b961918ee22fccdeab23532511d) and synchronized the doc, thanks for the reminder.

**Review by LujhCoconut — APPROVED** [review]
While it feels slightly odd to me to allocate based solely on SSD capacity — intuitively, factoring in DRAM capacity as well would seem more appropriate (e.g., `alpha * ssd_free + (1-alpha) * ddr_free`) — this PR is isolated to a single policy and doesn't affect other code paths, so I'm fine with merging it.

**Comment by zchuango** [comment]
Thanks for @LujhCoconut for the review and approval. 
All checks are passing, and there are no conflicts with the base branch.
Could the maintainers please take a final look when convenient?
@ykwd @stmatengss @XucSh

**Comment by zchuango** [comment]
> Thanks for @LujhCoconut for the review and approval. All checks are passing, and there are no conflicts with the base branch. Could the maintainers please take a final look when convenient? @ykwd @stmatengss @XucSh

Some CI test are not passing, since our commit haven't sync for some time, this is under fixing.

**Comment by zchuango** [comment]
Sorry for the delay. We did some major fixes (and also added some tests)to catch up with latest main branch, and all CI tests have passed. Could you please have another look when convenient? Thanks~ @LujhCoconut
